app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,200 @@
+    """Create a sample ingredient"""
+    default = {
+        'name': 'Test ingredient'
+    }
+    default.update(params)
+
+    return Ingredient.objects.create(recipe=recipe, **default)
+
+
+class TestRecipeApi(TestCase):
+    """Class to test the recipe API"""
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_recipe_list(self):
+        """Test the retrieving values on the list"""
+        sample_recipe(name='Recipe one')
+        sample_recipe(name='Recipe two')
+
+        res = self.client.get(RECIPE_URL)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(res.data), 2)
+
+    def test_recipe_filtering_list(self):
+        """Test tje retrieving values filtering by name"""
+        recipe_01 = sample_recipe(name='Recipe one')
+        recipe_02 = sample_recipe(name='Number two recipe')
+        recipe_03 = sample_recipe(name="Another one")
+
+        res = self.client.get(RECIPE_URL, {'name': 'rec'})
+
+        serialized_data_01 = RecipeSerializer(recipe_01)
+        serialized_data_02 = RecipeSerializer(recipe_02)
+        serialized_data_03 = RecipeSerializer(recipe_03)
+
+        self.assertIn(serialized_data_01.data, res.data)
+        self.assertIn(serialized_data_02.data, res.data)
+        self.assertNotIn(serialized_data_03.data, res.data)
+
+    def test_recipe_detail(self):
+        """Test the retrieving values of one recipe"""
+        recipe = sample_recipe()
+
+        url = recipe_detail_url(recipe.id)
+        res = self.client.get(url)
+
+        serializer = RecipeSerializer(recipe)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(serializer.data, res.data)
+
+    def test_recipe_valid_creation_without_ingredients(self):
+        """Test the valid creation of a recipe without ingredients"""
+        payload = {
+            'name': 'Scrambled eggs',
+            'description': 'Delicious and easy'
+        }
+
+        res = self.client.post(RECIPE_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        for key in payload.keys():
+            self.assertEqual(payload[key], getattr(recipe, key))
+
+    def test_recipe_valid_creation_with_ingredient(self):
+        """Test the valid creation adding ingredients since the beginning"""
+        payload = {
+            'name': 'Chicken over rice',
+            'description': 'Delicious plate',
+            'ingredients': [
+                {'name': 'chicken'},
+                {'name': 'rice'},
+                {'name': 'salt'}
+            ]
+        }
+
+        res = self.client.post(RECIPE_URL, payload, format='json')
+
+        recipe = Recipe.objects.get(id=res.data['id'])
+
+        ingredients = recipe.ingredients.all()
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        self.assertEqual(len(ingredients), 3)
```

Comment:
You could also do `ingredients.count()` instead of length. In this particular case it doesn't matter, but if Ingredients was a huge object and you had a couple of more, you'd retrieve them all from the DB just to count them. 
With `.count()` you issue just a `COUNT` query to the DB.

Reply:
Here was a mistake of mine. At the beginning I wanted to validate more things and I tried to avoid the continues database calls for each validation, but after I saw the count validation gonna be enough. I'm optimizing that part